### PR TITLE
WIP. getTile() with quad tree caching.

### PR DESCRIPTION
**Pull request by OleksiiZubko**
This CR fixes the issue with the fetching quad tree index
with depth 0.

Fetching quad tree index for parent Tile of the requested Tile
with delta 4 is useful for caching all sub quads
and parent quads metadata for the later calls.

Returns the blob of the requested Tile or the blob
of the closest parent Tile.

Also changed getData method in Versioned and Volatile layer clients.
The method caches quad tree index with delta 4
and returns data of the requested tile or rejects if now found.

Added getAggregatedData method to the Versioned and
Volatile layer clients.
The method re-uses getTile function to return data
of the requested tile or its closest ancestor.

Added the integration test for testing getting tile and caching.

Resolves: OLPEDGE-1953, OLPEDGE-1893

### CODE REVIEW

File: @here/olp-sdk-dataservice-read/lib/utils/getTile.ts [hunk 1]
```diff
@@ -138,13 +144,19 @@ export async function getTile(
...
     ) {
         if (fetchOptions !== FetchOptions.OnlineOnly) {
-            cache.put(partitionsRequest, catalogHrn.toString(), layerId, [
-                {
-                    partition: partitionId,
-                    dataHandle: quadTreeIndex.subQuads[0].dataHandle,
-                    version: quadTreeIndex.subQuads[0].version
-                }
-            ]);
+            const subQuadsToPartitionsList = quadTreeIndex.subQuads.map(
```
Comment (andescu) [inline @here/olp-sdk-dataservice-read/lib/utils/getTile.ts:147]: you need to also cache the parentQuads.

File: @here/olp-sdk-dataservice-read/lib/utils/getTile.ts [hunk 2]
```diff
@@ -115,7 +119,9 @@ export async function getTile(
         catalogHrn,
         layerId,
         layerType
-    ).withQuadKey(quadKey);
+    )
+        .withQuadKey(quadKey)
+        .withDepth(depth);
```
Comment (andescu) [inline @here/olp-sdk-dataservice-read/lib/utils/getTile.ts:124]: I think there is a misunderstanding here. If the user requests a tile, e.g. 1234556789 on level 14, you do not use this tile as quadKey base for the quad tree index call. You go to its parent on level 10 and request with that one with depth 4 so that you get all the neighbouring tiles of 1234556789 and the parents on level 13, the grandparents on level 12 and the grandgrandparents on level 11. If you request quad tree index with a level 14 tile with dept 4 you get 1 partition back.

So basically you need to do something like:

```suggestion
        .withQuadKey(computeParentKey(quadKey, depth))
        .withDepth(depth);
```

File: @here/olp-sdk-dataservice-read/lib/utils/getTile.ts [hunk 3]
```diff
@@ -138,13 +144,19 @@ export async function getTile(
...
         if (fetchOptions !== FetchOptions.OnlineOnly) {
-            cache.put(partitionsRequest, catalogHrn.toString(), layerId, [
-                {
-                    partition: partitionId,
-                    dataHandle: quadTreeIndex.subQuads[0].dataHandle,
-                    version: quadTreeIndex.subQuads[0].version
-                }
-            ]);
+            const subQuadsToPartitionsList = quadTreeIndex.subQuads.map(
+                item => ({
```
Comment (haraldF) [inline @here/olp-sdk-dataservice-read/lib/utils/getTile.ts:148]: What about sparse layers? If I read the code correctly, if we request a tile that doesn't exist, it fetches the index again?

File: @here/olp-sdk-dataservice-read/lib/utils/getTile.ts (hunk 1 above)
Comment (OleksiiZubko) [reply @here/olp-sdk-dataservice-read/lib/utils/getTile.ts:147]: I'm caching on 161 line

Comment (andescu) [reply @here/olp-sdk-dataservice-read/lib/utils/getTile.ts:147]: Sorry, missed that. Ignore this comment.

File: @here/olp-sdk-dataservice-read/lib/utils/getTile.ts (hunk 2 above)
Comment (OleksiiZubko) [reply @here/olp-sdk-dataservice-read/lib/utils/getTile.ts]: Updated, thanks. Now we're caching the Quad Tree from parent of the requested tile with delta 4 and returning the blob of the requested Tile or the blob of the closest parent Tile. (the same datahandle)

File: @here/olp-sdk-dataservice-read/lib/utils/getTile.ts (hunk 3 above)
Comment (OleksiiZubko) [reply @here/olp-sdk-dataservice-read/lib/utils/getTile.ts:152]: Updated the code. Returns the blob of the requested Tile or the blob of the closest parent Tile.
(or rejects with error for sure)

File: @here/olp-sdk-dataservice-read/lib/client/VersionedLayerClient.ts [hunk 4]
```diff
@@ -194,34 +222,41 @@ export class VersionedLayerClient {
...
-                          `No dataHandle for quadKey {column: ${quadKey.column}, row: ${quadKey.row}, level: ${quadKey.level}}. HRN: ${this.hrn}`
-                      )
-                  );
+            const subQuad = quadTreeIndexResponse.subQuads.find(
+                item => item.subQuadKey === mortonCode
+            );
+
+            if (!subQuad) {
+                return Promise.reject(noContentError);
+            }
```
Comment (mykhailo-kuchma) [inline @here/olp-sdk-dataservice-read/lib/client/VersionedLayerClient.ts:253]: What about checking the parent next? Note that subQuads array could be empty

Comment (OleksiiZubko) [reply @here/olp-sdk-dataservice-read/lib/client/VersionedLayerClient.ts:253]: it's per Andrey's request. Checking parents is in getAggregatedData only

Comment (OleksiiZubko) [reply @here/olp-sdk-dataservice-read/lib/client/VersionedLayerClient.ts:254]: "Note that subQuads array could be empty" in this case subQuad will be undefined

**Comment by OleksiiZubko** [comment]
Thanks for the reviews. I'm closing that CR and will make the smallest CRs for the better review process.
